### PR TITLE
[TIMOB-24593] Update Facebook Android SDK + iOS SDK to 4.25.0, support Graph v2.9 + FBSDKPlacesKit API, refactor iOS-project

### PR DESCRIPTION
**Pull request by hansemannn**
**JIRA**: https://jira.appcelerator.org/browse/TIMOB-24593

- [x] Android: [facebook-android-6.3.0.zip](https://github.com/appcelerator-modules/ti.facebook/files/1231819/facebook-android-6.3.0.zip)

- [x] iOS: [facebook-iphone-5.5.0.zip](https://github.com/appcelerator-modules/ti.facebook/files/1239432/facebook-iphone-5.5.0.zip)

### CODE REVIEW

**Comment by hansemannn** [comment]
@garymathews I finished moving the "old" jar-library to our new AAR-handling and updated to the latest FB-SDK. We would probably need to bump the minimum version of Ti.Facebook to 6.1.0 then, is that okay?

**Comment by garymathews** [comment]
@hansemannn Yes, the older module can be used if pre-6.1.0 compatibility is needed 👍 

I'll review this very soon.

**Review by garymathews — APPROVED** [review]
CR: **PASS**

File: ios/Classes/FacebookModule.m [hunk 1]
```diff
@@ -608,12 +622,29 @@ -(void)presentMessengerDialog:(id)args
...
     TiThreadPerformOnMainThread(^{
         FBSDKShareLinkContent *content = [[FBSDKShareLinkContent alloc] init];
+        
+        NSString *description = [params objectForKey:@"description"];
+        NSString *title = [params objectForKey:@"title"];
+        NSString *picture = [params objectForKey:@"picture"];
+        
+        if (description != nil) {
+            NSLog(@"Ti.Facebook.presentMessengerDialog.description has been deprecated in Ti.Facebook 5.5.0 as of the Graph v2.9 changes.");
+            content.contentDescription = [params objectForKey:@"description"];
```
Comment (vijaysingh-axway) [inline ios/Classes/FacebookModule.m]: As per FBSDKShareLinkContent.h - 
@property (nonatomic, readonly) NSString *contentDescription
  DEPRECATED_MSG_ATTRIBUTE("`contentDescription` is deprecated from Graph API 2.9");

contentDescription  has changed to read only, but you are assigning to it. It is giving build error. Please remove it.

File: ios/Classes/FacebookModule.m [hunk 2]
```diff
@@ -608,12 +622,29 @@ -(void)presentMessengerDialog:(id)args
...
+        }
+        
+        if (title != nil) {
+            NSLog(@"Ti.Facebook.presentMessengerDialog.title has been deprecated in Ti.Facebook 5.5.0 as of the Graph v2.9 changes.");
+            content.contentTitle = title;
```
Comment (vijaysingh-axway) [inline ios/Classes/FacebookModule.m]: @property (nonatomic, readonly) NSString *contentTitle
  DEPRECATED_MSG_ATTRIBUTE("`contentTitle` is deprecated from Graph API 2.9");

contentTitle  has changed to read only, but you are assigning to it. It is giving build error. Please remove it.

File: ios/Classes/FacebookModule.m [hunk 3]
```diff
@@ -608,12 +622,29 @@ -(void)presentMessengerDialog:(id)args
...
+        }
+        
+        if (picture != nil) {
+            NSLog(@"Ti.Facebook.presentMessengerDialog.picture has been deprecated in Ti.Facebook 5.5.0 as of the Graph v2.9 changes.");
+            content.imageURL = [NSURL URLWithString:picture];
```
Comment (vijaysingh-axway) [inline ios/Classes/FacebookModule.m]: As per FBSDKShareLinkContent.h -

@property (nonatomic, readonly) NSURL *imageURL
  DEPRECATED_MSG_ATTRIBUTE("`imageURL` is deprecated from Graph API 2.9");
imageURL has changed to read only, but you are assigning to it. It is giving build error. Please remove it.

File: ios/Classes/FacebookModule.m [hunk 4]
```diff
@@ -574,17 +580,25 @@ -(void)presentShareDialog:(id)args
...
-        if ([params objectForKey:@"name"] != nil) {
-            DEPRECATED_REPLACED_REMOVED(@"Facebook.presentShareDialog.name", @"5.0.0", @"5.0.0", @"Titanium.Facebook.presentShareDialog.title");
+        
+        NSString *description = [params objectForKey:@"description"];
+        NSString *title = [params objectForKey:@"title"];
+        NSString *picture = [params objectForKey:@"picture"];
+        
+        if (description != nil) {
+            NSLog(@"Ti.Facebook.presentShareDialog.description has been deprecated in Ti.Facebook 5.5.0 as of the Graph v2.9 changes.");
+            content.contentDescription = [params objectForKey:@"description"];
```
Comment (vijaysingh-axway) [inline ios/Classes/FacebookModule.m]: As per FBSDKShareLinkContent.h -
@property (nonatomic, readonly) NSString *contentDescription
DEPRECATED_MSG_ATTRIBUTE("contentDescription is deprecated from Graph API 2.9");

contentDescription has changed to read only, but you are assigning to it. It is giving build error. Please remove it.

File: ios/Classes/FacebookModule.m [hunk 5]
```diff
@@ -574,17 +580,25 @@ -(void)presentShareDialog:(id)args
...
         }
-        if ([params objectForKey:@"title"] != nil){
-            content.contentTitle = [params objectForKey:@"title"];
+ 
+        if (title != nil) {
+            NSLog(@"Ti.Facebook.presentShareDialog.title has been deprecated in Ti.Facebook 5.5.0 as of the Graph v2.9 changes.");
+            content.contentTitle = title;
```
Comment (vijaysingh-axway) [inline ios/Classes/FacebookModule.m]: @property (nonatomic, readonly) NSString *contentTitle
DEPRECATED_MSG_ATTRIBUTE("contentTitle is deprecated from Graph API 2.9");

contentTitle has changed to read only, but you are assigning to it. It is giving build error. Please remove it.

File: ios/Classes/FacebookModule.m [hunk 6]
```diff
@@ -574,17 +580,25 @@ -(void)presentShareDialog:(id)args
...
         }
-        if ([params objectForKey:@"caption"] != nil) {
-            DEPRECATED_REMOVED(@"Facebook.presentShareDialog.caption", @"5.0.0", @"5.0.0");
+        
+        if (picture != nil) {
+            NSLog(@"Ti.Facebook.presentShareDialog.picture has been deprecated in Ti.Facebook 5.5.0 as of the Graph v2.9 changes.");
+            content.imageURL = [NSURL URLWithString:picture];
```
Comment (vijaysingh-axway) [inline ios/Classes/FacebookModule.m]: As per FBSDKShareLinkContent.h -
@property (nonatomic, readonly) NSURL *imageURL
DEPRECATED_MSG_ATTRIBUTE("imageURL is deprecated from Graph API 2.9");
imageURL has changed to read only, but you are assigning to it. It is giving build error. Please remove it.

File: apidoc/Facebook.yml [hunk 7]
```diff
@@ -1913,36 +1928,37 @@ summary: |
...
-        removed: "5.0.0"
-
   - name: title
     summary: Title associated with the link.
     type: String
     since: "5.0.0"
+    deprecated:
+        since: "6.1.0"
+        notes: |
+            Facebook deprecated the "description" key in Graph v2.9, it will be
```
Comment (vijaysingh-axway) [inline apidoc/Facebook.yml]: Rather "description" it should be "title".

File: apidoc/Facebook.yml [hunk 8]
```diff
@@ -1913,36 +1928,37 @@ summary: |
...
+            removed in the future.
 
   - name: picture
     summary: Link to a thumbnail to associate with the link.
     type: String
+    deprecated:
+        since: "6.1.0"
+        notes: |
+            Facebook deprecated the "description" key in Graph v2.9, it will be
```
Comment (vijaysingh-axway) [inline apidoc/Facebook.yml]: Rather "description" it should be "picture".

**Comment by hansemannn** [comment]
@vijaysingh-axway Thanks for the feedback, I've eliminated all of them!

**Comment by hansemannn** [comment]
Hey @vijaysingh-axway, this unintentionally ended in basically rewriting the whole module. This includes:
- Moved to ARC
- Exposed public API's to headers
- Added nullability-specification
- Rename parameter to human-readable ones

**Review by vijaysingh-axway — APPROVED** [review]
CR passed.

**Comment by hansemannn** [comment]
@garymathews Double-checking: Can we merge this? iOS is ready, Android seems to be approved by you as well. Thanksss!

**Review by garymathews — APPROVED** [review]
CR: **PASS**

**Comment by hansemannn** [comment]
@ewieberappc Can we do a FT before merging here? I feel we are more save to test the packaged modules before. It's provided above!

**Comment by hansemannn** [comment]
Side-note: CI fails because it tries with master, which requires the V8 upgrade for Android, so not related to this PR.

**Review by mukherjee2 — APPROVED** [review]
Node Version: 6.10.3
NPM Version: 3.10.10
Mac OS: 10.12.4
Appc CLI: 6.2.3
Appc CLI NPM: 4.2.9
Titanium SDK version: 6.2.0.v20170823150008
Appcelerator Studio, build: 4.9.1.201707200100
Xcode 8.3.2
Ti Facebook Module 6.3.0 for Android

FR passed for Android. I used the latest module, and was able to use the module with a Facebook app that I created.

**Pull request merged**